modules/ml/src/main/java/org/apache/ignite/ml/recommendation/RecommendationTrainer.java

```diff
@@ -0,0 +1,319 @@
+/*
+ * Copyright 2019 GridGain Systems, Inc. and Contributors.
+ *
+ * Licensed under the GridGain Community Edition License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.gridgain.com/products/software/community-edition/gridgain-community-edition-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.ml.recommendation;
+
+import java.io.Serializable;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Random;
+import java.util.Set;
+import org.apache.ignite.ml.dataset.Dataset;
+import org.apache.ignite.ml.dataset.DatasetBuilder;
+import org.apache.ignite.ml.dataset.primitive.builder.context.EmptyContextBuilder;
+import org.apache.ignite.ml.dataset.primitive.context.EmptyContext;
+import org.apache.ignite.ml.environment.LearningEnvironment;
+import org.apache.ignite.ml.environment.LearningEnvironmentBuilder;
+import org.apache.ignite.ml.math.primitives.vector.Vector;
+import org.apache.ignite.ml.math.primitives.vector.VectorUtils;
+import org.apache.ignite.ml.recommendation.util.MatrixFactorizationGradient;
+import org.apache.ignite.ml.recommendation.util.RecommendationDatasetData;
+import org.apache.ignite.ml.recommendation.util.RecommendationDatasetDataBuilder;
+
+/**
+ * Trainer of the recommendation system.
+ */
+public class RecommendationTrainer {
+    /** Environment builder. */
+    private LearningEnvironmentBuilder environmentBuilder = LearningEnvironmentBuilder.defaultBuilder();
+
+    /** Trainer learning environment. */
+    private LearningEnvironment trainerEnvironment = environmentBuilder.buildForTrainer();
+
+    /** Batch size of stochastic gradient descent. The size of a dataset used on each step of SGD. */
+    private int batchSize = 1000;
+
+    /** Regularizer. */
+    private double regularizer = 0.0;
```

Comment:
regularizer is function that try to setup model with some restrictions. This is the strength parameter of regularizer

Reply:
Done.